src/lib/mongoBackend/mongoConnectionPool.cpp

```diff
@@ -0,0 +1,244 @@
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero
+* General Public License for more details.
+*
+* You should have received a copy of the GNU Affero General Public License
+* along with Orion Context Broker. If not, see http://www.gnu.org/licenses/.
+*
+* For those usages not covered by this license please contact with
+* iot_support at tid dot es
+*
+* Author: Ken Zangelin
+*/
+#include <time.h>
+#include <semaphore.h>
+
+#include "logMsg/logMsg.h"
+#include "common/clockFunctions.h"
+#include "mongoBackend/mongoConnectionPool.h"
+#include "mongoBackend/MongoGlobal.h"
+
+
+
+/* ****************************************************************************
+*
+* MongoConnection - 
+*/
+typedef struct MongoConnection
+{
+  DBClientBase*  connection;
+  bool           free;
+} MongoConnection;
+
+
+
+/* ****************************************************************************
+*
+* globals - 
+*/
+static MongoConnection* connectionPool     = NULL;
+static int              connectionPoolSize = 0;
+static sem_t            connectionPoolSem;
+static sem_t            connectionSem;
+static struct timespec  semWaitingTime     = { 0, 0 };
+static bool             semStatistics      = false;
+
+
+
+/* ****************************************************************************
+*
+* mongoConnectionPoolInit - 
+*/
+int mongoConnectionPoolInit
+(
+  const char* host,
+  const char* db,
+  const char* rplSet,
+  const char* username,
+  const char* passwd,
+  double      timeout,
+  int         writeConcern,
+  int         poolSize,
+  bool        semTimeStat
+)
+{
+  //
+  // Checking incoming parameters
+  //
+  if (poolSize == 0)
+  {
+    poolSize = 10;
+  }
+
+
+  //
+  // Create the pool
+  //
+  connectionPool     = (MongoConnection*) calloc(sizeof(MongoConnection), poolSize);
+  if (connectionPool == NULL)
+  {
+    LM_E(("Runtime Error (insufficient memory to create connection pool of %d connections)", poolSize));
+    return -1;
+  }
+  connectionPoolSize = poolSize;
+
+  //
+  // Initialize (connect) the pool
+  //
+  for (int ix = 0; ix < connectionPoolSize; ++ix)
+  {
+    connectionPool[ix].connection = mongoConnect(host, db, rplSet, username, passwd, writeConcern, timeout);
+    connectionPool[ix].free       = true;
+  }
+
+  //
+  // Set up the semaphore protecting the pool itself (connectionPoolSem)
+  //
+  int r = sem_init(&connectionPoolSem, 0, 1);
+
+  if (r != 0)
+  {
+    LM_E(("Runtime Error (cannot create connection pool semaphore)"));
+    return -1;
+  }
+
+  //
+  // Set up the semaphore protecting the set of connections of the pool (connectionSem)
+  // Note that this is a counting semaphore, initialized to connectionPoolSize.
+  //
+  r = sem_init(&connectionSem, 0, connectionPoolSize);
+  if (r != 0)
+  {
+    LM_E(("Runtime Error (cannot create connection semaphore-set)"));
+    return -1;
+  }
+
+  // Measure accumulated semaphore waiting time?
```

Comment:
The question mark as measure or not measure depend on the parameter semTimeStat (on the next line).